src/java/org/apache/cassandra/index/sai/utils/TypeUtil.java

```diff
@@ -130,6 +131,23 @@ public static ByteComparable max(ByteComparable a, ByteComparable b)
         return a == null ? b : (b == null || ByteComparable.compare(b, a, ByteComparable.Version.OSS50) < 0) ? a : b;
     }
 
+    /**
+     * Returns the value length for the given {@link AbstractType}, selecting 16 for types
+     * that officially use VARIABLE_LENGTH but are, in fact, of a fixed length.
+     */
+    public static int fixedSizeOf(AbstractType<?> type)
```

Comment:
Not directly related to this ticket, but it seems to me that the methods on `TypeUtil` do a lot of repeated work in the hot path just to determine the data type of the caller. These operations include finding the base type of reversed multiple times, multiple `instanceof` calls, etc. All these operations on each index are called with the same `IndexContext#getValidator` argument, needlessly repeating work for every column value.

I understand that most of the methods on `TypeUtil` are things that would normally be part of `AbstractType`, so each particular data type can provide its own implementation. But we don't want to couple the generic data types with SAI, so we have this class instead.

I think that we could add a kind of `TermType` class decorating `AbstractType`, and subclass it for every data type that gets special treatment (`LONG`, `VARINT`, `DECIMAL` and `INET`). Then, `IndexContext` would hold the adequate instance of `TermType`. This instance would be able to provide the same operations as the current `TypeUtil`, but without the type checks.

If that makes sense, we could do it on a separate ticket.

Reply:
A separate ticket makes sense to me. After all, this ticket doesn't introduce the `TypeUtil` class.